### ISSUE TITLE
Keras to tflite. Multiple outputs all given name Identity

### ISSUE THREAD

**Issue by marno1d** [tensorflow/tensorflow#32180]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below):2.0-rc0
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.0/7.6
- GPU model and memory: RTX 2070

**Describe the current behavior**
When I convert a keras model to a tflite model all outputs are named Identity. When you have a bunch of outputs that all have the same output size it becomes impossible to tell which one is which.

**Describe the expected behavior**
I would expect the names to match the layer names that the output comes from. 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
import tensorflow as tf

inputs = []
outputs = []

x = tf.keras.layers.Input(shape=(10), batch_size=1, name='input1')
inputs.append(x)
x = tf.keras.layers.Dense(10, name='dense1')(x)
outputs.append(x)
x = tf.keras.layers.Dense(10, name='dense2')(x)
outputs.append(x)

model = tf.keras.Model(inputs=inputs, outputs=outputs)

model.summary()
```
```
Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input1 (InputLayer)          [(1, 10)]                 0         
_________________________________________________________________
dense1 (Dense)               (1, 10)                   110       
_________________________________________________________________
dense2 (Dense)               (1, 10)                   110       
=================================================================
Total params: 220
Trainable params: 220
Non-trainable params: 0
```
```
model.outputs
```
```
[<tf.Tensor 'dense1/Identity:0' shape=(1, 10) dtype=float32>,
 <tf.Tensor 'dense2/Identity:0' shape=(1, 10) dtype=float32>]
```
```
# Convert and save the model.
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()
open("./save/simple.tflite", "wb").write(tflite_model)

# Load TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_path="./save/simple.tflite")
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

output_details
```
```
[{'name': 'Identity',
  'index': 0,
  'shape': array([ 1, 10], dtype=int32),
  'dtype': numpy.float32,
  'quantization': (0.0, 0)},
 {'name': 'Identity_1',
  'index': 1,
  'shape': array([ 1, 10], dtype=int32),
  'dtype': numpy.float32,
  'quantization': (0.0, 0)}]
```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by imskr** [comment]
I would like to work on this issue!  Please assign

**Comment by Cospel** [comment]
Thumbs up! Right now it is impossible to deploy models to tflite if they have multiple outputs. Is anybody working on this?

**Comment by marno1d** [comment]
@Cospel One work around I found was using https://lutzroeder.github.io/netron/ to view the tflite model to visually figure out what index each output corresponded to. This issue should still be fixed but this work around should help people that are currently stuck on this problem.

**Comment by Cospel** [comment]
@marno1d Thank you for the idea. However this is not very friendly and scalable solution :).

I hoped that creating tf.keras.layers.Lambda with the name would work:

    output1 = tf.keras.layers.Lambda(lambda k: k, name='output1')(x)

but it's just creates another Identity....

**Comment by marno1d** [comment]
@Cospel I agree that it isn't a scalable solution. I think all outputs are automatically wrapped in an Identity layer according to the source code. I'm not familiar enough with the code to know why this is there / how to fix it. 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/base_layer.py#L774

`# Wrap Tensors in "outputs" in "tf.identity" to avoid`
 ` # circular dependencies.`
 ` outputs = base_layer_utils.mark_as_return(outputs, acd)`

**Comment by unexge** [comment]
@marno1d yes, making a change like this fixes the issue (and probably creates another issues 😄)

https://github.com/tensorflow/tensorflow/blob/v2.1.0-rc2/tensorflow/python/framework/auto_control_deps.py#L167
```diff
-tensor = array_ops.identity(tensor)
+# convert "model/tf_op_layer_output/output:0" -> "output"
+tensor = array_ops.identity(tensor, tensor.name.split("/")[-1].split(":")[0])
```

**Comment by ghop02** [comment]
I'm also running into this. 

@ravikyram or @jdduke have you seen this? Don't have an easy way of getting around it right now

**Comment by ghop02** [comment]
okay, so it seems that at some point, it sorts the output tensors by name. So, you're able to control the outputs by prepending the outputs with a letter:
```
x = tf.keras.layers.Input(shape=(10), batch_size=1, name='input1')
y = tf.keras.layers.Dense(10, name='a_dense1')(x)
z = tf.keras.layers.Dense(20, name='b_dense2')(x)
model = tf.keras.Model(x, [y, z])

converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()
path = pathlib.Path('test_order1.tflite')
path.write_bytes(tflite_model)

# Names flipped
x = tf.keras.layers.Input(shape=(10), batch_size=1, name='input1')
y = tf.keras.layers.Dense(10, name='b_dense1')(x)
z = tf.keras.layers.Dense(20, name='a_dense2')(x)
model = tf.keras.Model(x, [y, z])

converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()
path = pathlib.Path('test_order2.tflite')
path.write_bytes(tflite_model)
```
### tflite_order1.tflite
![image](https://user-images.githubusercontent.com/1422280/72569644-e2342f80-3888-11ea-8592-5995091297a8.png)

### tflite_order2.tflite
![image](https://user-images.githubusercontent.com/1422280/72569663-efe9b500-3888-11ea-971a-3443d195301e.png)


This obviously is not great, but a temporary bandaid that works for me

**Comment by kmkolasinski** [comment]
Is anybody working on this ? This problem still appears in the newer version of TF.

**Comment by kmkolasinski** [comment]
Hi, I have found (temporal) solution for this problem. It seems that the invalid output names are generated within this function:

```python
tensorflow_core.python.framework.func_graph.func_graph_from_py_func
```
It has a default argument `add_control_dependencies` set to True. Switching it to False before model conversion leads to correct names when combined with keras.Lambda and tf.identity e.g:

```python
output = Lambda(lambda x: tf.identity(x, name="output"), name="logits")(outputs)
keras.Model(input_image, output, name="export")
```
gives me following output details:
```
Output details:
{ 'dtype': <class 'numpy.float32'>,
  'index': 192,
  'name': 'export/logits/output',
  'quantization': (0.0, 0),
  'shape': array([  1, 4], dtype=int32)}
```
 
My shitfix for this is to use Monkey patching and create `tflite_converter_shitfix.py` file which will set `add_control_dependencies` to False at runtime just before model conversion:

```python
import weakref
import tensorflow as tf
from tensorflow.python.eager import def_function as _def_function
from tensorflow.python.eager import function as _function
from tensorflow.python.framework import func_graph as func_graph_module
from tensorflow.python.framework import ops
from tensorflow.python.keras.saving import saving_utils as _saving_utils
from tensorflow.python.util import tf_decorator


keras = tf.keras


def from_keras_model(model: keras.Model) -> tf.lite.TFLiteConverter:
    """Creates a TFLiteConverter object from a Keras model.

    Args:
      model: tf.Keras.Model

    Returns:
      TFLiteConverter object.
    """
    input_signature = None
    # If the model's call is not a `tf.function`, then we need to first get its
    # input signature from `model_input_signature` method. We can't directly
    # call `trace_model_call` because otherwise the batch dimension is set
    # to None.
    # Once we have better support for dynamic shapes, we can remove this.
    if not isinstance(model.call, _def_function.Function):
        # Pass `keep_original_batch_size=True` will ensure that we get an input
        # signature including the batch dimension specified by the user.
        input_signature = _saving_utils.model_input_signature(
            model, keep_original_batch_size=True
        )

    func = _saving_utils.trace_model_call(model, input_signature)
    func._defun_with_scope = lambda s: _defun_with_scope(func, s)
    concrete_func = func.get_concrete_function()
    return tf.lite.TFLiteConverter([concrete_func])


def _defun_with_scope(self, scope):
    """Creates a defun wrapped inside a variable creator scope."""

    weak_wrapped_fn = None

    def wrapped_fn(*args, **kwds):
        """Wraps `self._python_function` in a variable creator scope."""
        with ops.get_default_graph()._variable_creator_scope(scope, priority=50):
            return weak_wrapped_fn().__wrapped__(*args, **kwds)

    weak_wrapped_fn = weakref.ref(wrapped_fn)

    fun = self._defun(tf_decorator.make_decorator(self._python_function, wrapped_fn))

    fun._create_graph_function = lambda args, kwargs: _create_graph_function(
        fun, args, kwargs
    )
    return fun


def _create_graph_function(self, args, kwargs, override_flat_arg_shapes=None):
    """Create a `ConcreteFunction` from `args` and `kwargs`."""

    self.tracing_count += 1
    if self.input_signature is None:
        arglen = len(args)
    else:
        arglen = len(self.input_signature)
    base_arg_names = self._function_spec.arg_names[:arglen]
    num_missing_args = arglen - len(self._function_spec.arg_names)
    missing_arg_names = [self._function_spec.vararg_name] * num_missing_args
    # Produce a list of missing args of the form ["arg_0", "arg_1", ...],
    # where arg is based on the self._function_spec.vararg_name.
    missing_arg_names = ["%s_%d" % (arg, i) for i, arg in enumerate(missing_arg_names)]
    arg_names = base_arg_names + missing_arg_names

    graph_function = _function.ConcreteFunction(
        func_graph_module.func_graph_from_py_func(
            self._name,
            self._python_function,
            args,
            kwargs,
            self.input_signature,
            autograph=self._autograph,
            autograph_options=self._autograph_options,
            arg_names=arg_names,
            override_flat_arg_shapes=override_flat_arg_shapes,
            capture_by_value=self._capture_by_value,
            add_control_dependencies=False,
        ),
        self._function_attributes,
        # Tell the ConcreteFunction to clean up its graph once it goes out of
        # scope. This is not the default behavior since it gets used in some
        # places (like Keras) where the FuncGraph lives longer than the
        # ConcreteFunction.
        shared_func_graph=False,
    )
    return graph_function
```

And then use:

```python

from tflite_converter_shitfix import from_keras_model
converter = from_keras_model(model)
converter.experimental_new_converter = True  
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.convert()
```

**Comment by krille90** [comment]
Hi, I've run into this today and figured out that I can still use the old converter code with disabled eager execution:

```
tf.compat.v1.disable_eager_execution()

converter = tf.compat.v1.lite.TFLiteConverter.from_keras_model_file(path_model)

return converter.convert()
```

this will fix it for my use case. I hope it helps others too.

**Comment by donbekci-neosensory** [comment]
The problem persists in TF2.3 nightly as well.

**Comment by Lannister-Xiaolin** [comment]
I also encountered this problem。

**Comment by rtapps** [comment]
This also happens on tensorrt conversion.
I tracked down the cause to 'canonicalize_signatures' method in signature_serialization.py
Seems like only the inputs are being saved in the signature and the outputs are being converted to "Identity" named tensors.

**Comment by JiaqingFu** [comment]
I run into the problem too。is it solved？ tflite for multiple output

**Comment by LIFEPANTS** [comment]
Same issue here, All outputs ordered after their defined names but shows as Identity

**Comment by caegomezji** [comment]
I got this also...  all my outpus were re-ordered

**Comment by abattery** [comment]
@caegomezji 

SignatureDef provides meaningful/generic names for inputs/outputs which doesn't rely on specific model details. More on SignatureDefs [here](https://www.tensorflow.org/guide/saved_model?hl=zh-tw#specifying_signatures_during_export).
If your saved model has a defined signatureDef then it will be exported during conversion to TFLite and then you can use the Signature inputs/outputs for inference and not relying on inputs/outputs order or tensor names.

SignatureDef support is new thing available in nightly and will be available in 2.5

Then use the nightly for converting to tflite. The generated tflite file will have the SignatureDef details.
Using Python API (as an example) you can do something like

```
my_signature = interpreter.get_signature_runner("my_method")
results = my_signature(input_1=input_tensor_1, input_2=input_tensor_2)
print(results["my_output"])
```

**Comment by maxpv** [comment]
Any update on this issue?

It is possible to rename a TF model's outputs for using `coremltools` but not `tensorflow?

**Comment by kdonbekci** [comment]
Looking forward to updates on this issue! For extreme use cases, every byte counts and would be great if we had the option to remove some of the unnecessary metadata and tensor names from the exported C array.

**Comment by abattery** [comment]
@kdonbekci @maxpv 

Please see https://github.com/tensorflow/tensorflow/issues/32180#issuecomment-772140542 for the solution.

**Comment by n-moussa** [comment]
@abattery 

If I'm not wrong, the issue is not solved at C++ level ?
Is there any way to solve it with tflite C++ API ?

Thanks

**Comment by karimnosseir** [comment]
@n-moussa There is a C++ API for using Signatures

* To get list of signatures available
interpreter.signature_def_names();
* To get inputs/output of a single Signature:
TfLiteTensor* input_1_tensor = interpreter.input_tensor_by_signature_name("input1", "mySignature");
TfLiteTensor* input_2_tensor = interpreter.input_tensor_by_signature_name("input2", "mySignature");
const TfLiteTensor* output_tensor = interpreter.output_tensor_by_signature_name("output", "mySignature");

Example,
interpreter.AllocateTensors()

autp* input1 = interpreter.input_tensor_by_signature_name("input1", "mySignature");
// Set the data in the tensor normally e,g,: input1->data.f[0] = 1.0f;
autp* input2 = interpreter.input_tensor_by_signature_name("input2", "mySignature");
// Set the data in the tensor normally e,g,: input1->data.f[0] = 1.0f;

// Run
interpreter.Invoke();

// Fetch outputs
const TfLiteTensor* output_tensor = interpreter.output_tensor_by_signature_name("output", "mySignature");

**Comment by LIFEPANTS** [comment]
Maybe i'm missing something obvious here but I don't understand how this was fixed? I still have the same problem with nightly 2.5.0-dev20210307
my signature def is intact in the saved model. and lists my outputs correctly

['serving_default']
{'myoutputname1': TensorSpec(shape=(None, 8), dtype=tf.float32, name='myoutputname1'), 'myoutputname2': TensorSpec(shape=(None, 1), dtype=tf.float32, name='myoutputname2'), 'myoutputname3': TensorSpec(shape=(None, 2), dtype=tf.float32, name='myoutputname3')}

then during conversion the serving_default is ignored and i get generic names:

name: serving_default_input_1:0
shape: [ 1 80 80  1]
type: <class 'numpy.float32'>
quantization_parameters: (0.0, 0)
0= 0.0
255= 0.0

== Output details ==
name: StatefulPartitionedCall:2
shape: [1 8]
type: <class 'numpy.float32'>
quantization_parameters: (0.0, 0)
0= 0.0
255= 0.0

== Output details ==
name: StatefulPartitionedCall:1
shape: [1 1]
type: <class 'numpy.float32'>
quantization_parameters: (0.0, 0)
0= 0.0
255= 0.0

== Output details ==
name: StatefulPartitionedCall:0
shape: [1 2]
type: <class 'numpy.float32'>
quantization_parameters: (0.0, 0)
0= 0.0
255= 0.0

**Comment by avroshk** [comment]
Love the signature runner API! I noticed that when I convert a `concrete_function` instead of a `saved_model` to tflite using `tf.lite.TFLiteConverter.from_concrete_functions()` the signature def is missing from the model after I load it into the interpreter. Is this behaviour by design or am I missing something during the conversion (for example, some attributes that I need to add to the concrete function, so it persists in the signature def)?   

Sorry for adding to a closed thread. I thought this was relevant to the discussion here.

**Comment by abattery** [comment]
@avroshk Thanks for your comment! Could you file another issue separately for the concrete function's signature def support? This can be a good one for the feature request. :-)

**Comment by karimnosseir** [comment]
Currently the Signature is supported through saved model. As Jaesung said please file another issue for a feature request.

Thanks

**Comment by manp-git** [comment]
Another this I realised is the weights and biases are getting modified in this procedure and graph structure is also changing when visualised through netron and the same changed structure is reflecting when checked via scripting to  cross validate the netron output. Conv2_block1_2 for example had no BN layer before Relu. But If want to pull out the Relu output then it adds BN layers at every node which is being pulled out

**Comment by karimnosseir** [comment]
@manp-git I am not sure i understand what you refer to by "this procedure"
Please file separate issue with sample and reproduce steps if you're having an issue.

Thanks

**Comment by LIFEPANTS** [comment]
this issue is partially back in the latest latest nightly and the official tf2.5.0
it works in nightly 2.5.0-dev20210307

When doing a straight conversion to tflite it works as expected, with input and output naming carried over as it should

But when quantizing a model through converter.optimizations = [tf.lite.Optimize.DEFAULT]
the signature is empty

**Comment by karimnosseir** [comment]
@LIFEPANTS Thanks for the report. We are working on a fix for signature getting dropped. It is work in progress now.

Sorry for the trouble.

**Comment by karimnosseir** [comment]
@MeghnaNatraj FYI

**Comment by LIFEPANTS** [comment]
@karimnosseir 
Was this ever fixed? any release i should try?

**Comment by abattery** [comment]
You can try out the nightly version. See also https://www.tensorflow.org/lite/guide/signatures

**Comment by SunseKwon** [comment]
> @caegomezji
> 
> SignatureDef provides meaningful/generic names for inputs/outputs which doesn't rely on specific model details. More on SignatureDefs [here](https://www.tensorflow.org/guide/saved_model?hl=zh-tw#specifying_signatures_during_export). If your saved model has a defined signatureDef then it will be exported during conversion to TFLite and then you can use the Signature inputs/outputs for inference and not relying on inputs/outputs order or tensor names.
> 
> SignatureDef support is new thing available in nightly and will be available in 2.5
> 
> Then use the nightly for converting to tflite. The generated tflite file will have the SignatureDef details. Using Python API (as an example) you can do something like
> 
> ```
> my_signature = interpreter.get_signature_runner("my_method")
> results = my_signature(input_1=input_tensor_1, input_2=input_tensor_2)
> print(results["my_output"])
> ```

you saved my life !!! so cool. thanks!!!
